functionary/sglang_inference.py

```diff
@@ -0,0 +1,641 @@
+        prompt = (
+            params.adapted_request.text
+            if params.adapted_request.text
+            else params.tokenizer.decode(params.adapted_request.input_ids)
+        )
+        # Iterates over the text generated by the SGLang Frontend Runtime
+        for out in params.frontend_state.text_iter():
+            if out.startswith(prompt):
+                continue
+            yield out, None
+        yield "", "stop"
+    else:
+        # Iterates over the text generated by the tokenizer manager
+        stream_buffer = ""
+        async for content in params.tokenizer_manager.generate_request(
+            params.adapted_request, params.raw_request
+        ):
+            text = content["text"]
+            delta = text[len(stream_buffer) :]
+            stream_buffer = stream_buffer + delta
+            finish_reason = content["meta_info"]["finish_reason"]
+
+            # If finish_reason is not None and delta_text is not empty,
+            # the delta_text is the eos_token and just remove it
+            if finish_reason is not None:
+                finish_reason = finish_reason["type"]
+                if len(delta) > 0:
+                    delta = ""
+            yield delta, finish_reason
+
+
+async def completion_stream_generator(params: ChatCompletionParams):
+    """
+    This asynchronous generator function produces a stream of ChatCompletionChunk
+    objects. It handles both grammar-sampling and regular generations,
+    depending on the parameters provided.
+
+    Args:
+        params (ChatCompletionParams): A dataclass containing all necessary
+            parameters for the chat completion, including the request details,
+            tokenizer, backend, and other configuration options.
+
+    Yields:
+        str: JSON-formatted strings representing chunks of the chat completion
+             response, including delta updates and finish reasons.
+
+    Notes:
+        - The function adapts its behavior based on whether grammar sampling
+          is enabled or not.
+        - It handles the conversion of tool calls to function calls when
+          appropriate.
+        - The stream is terminated with a "[DONE]" message.
+    """
+    # Initialize the text generator
+    generator = wrap_sgl_generator(params)
+
+    tool_call_count = 0
+    # Generate the text in openai format
+    async for response in generate_openai_format_from_stream_async(
+        generator,
+        params.prompt_template,
+        params.tool_func_choice,
+        params.tools_or_functions,
+    ):
+        # Convert tool_calls to function_call if request.functions is provided
+        if (
+            params.request.functions
+            and len(params.request.functions) > 0
+            and "tool_calls" in response["delta"]
+            and response["delta"]["tool_calls"]
+            and len(response["delta"]["tool_calls"]) > 0
+        ):
+            tool_name = response["delta"]["tool_calls"][0]["function"]["name"]
```

Comment:
I think this part is also in vllm, can we somehow implement a function in inference_utils.py then use this function in both vllm_inference and sglang_inference ?

Reply:
Fixed